lib/tail-file.js

```diff
@@ -0,0 +1,279 @@
+      } else if (this[kInode] !== stats.ino) {
+        // File renamed/rolled
+        await this._readRemainderFromFileHandle()
+        await this._openFile()
+        this[kStartPos] = 0
+        this[kInode] = stats.ino
+        fileHasChanged = true
+        this.emit('renamed', {
+          message: 'The file was renamed or rolled.  Tailing resumed from the beginning.'
+        , filename: this[kFileName]
+        , when: new Date()
+        })
+      } else if (eof < this[kStartPos]) {
+        // Same file, but smaller/truncated
+        this[kStartPos] = 0
+        this[kInode] = stats.ino
+        fileHasChanged = true
+        this.emit('truncated', {
+          message: 'The file was truncated.  Tailing resumed from the beginning.'
+        , filename: this[kFileName]
+        , when: new Date()
+        })
+      } else if (this[kStartPos] !== eof) {
+        fileHasChanged = true
+      }
+
+      if (fileHasChanged) {
+        await this._streamFileChanges()
+      } else {
+        setImmediate(this.emit.bind(this), 'flush')
+      }
+
+      this._scheduleTimer(this[kPollFileIntervalMs])
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        if (this[kFileHandle]) {
+          // Don't lose the last lines in the file if it previously existed.
+          // Perhaps it has not been re-created yet (or won't be)
+          try {
+            await this._readRemainderFromFileHandle()
+          } catch (error) {
+            this.emit('tail_error', error)
+          }
+        }
+        this[kPollFailureCount]++
+        if (this[kPollFailureCount] >= this[kMaxPollFailures]) {
+          return this.quit(err)
+        }
+        this.emit('retry', {
+          message: 'File disappeared. Retrying.'
+        , filename: this[kFileName]
+        , attempts: this[kPollFailureCount]
+        , when: new Date()
+        })
+        this._scheduleTimer(this[kPollFailureRetryMs])
+        return
+      }
+      // Some other error like EACCES
+      // TODO: Retries for certain error codes can be put here
+      return this.quit(err)
+    }
+    return
+  }
+
+  _scheduleTimer(ms) {
+    clearTimeout(this[kPollTimer])
+    if (this[kQuitting]) return
+    this[kPollTimer] = setTimeout(this._pollFileForChanges.bind(this), ms)
+    return
+  }
+
+  async _streamFileChanges() {
+    try {
+      const stream = fs.createReadStream(this[kFileName], {
+        ...this[kOpts].readStreamOpts
+      , start: this[kStartPos]
+      })
+      await this._readChunks(stream)
+    } catch (err) {
+      // Possible file removal.  Let auto-retry handle it.
+      this[kPollFailureCount]++
+      const error = new Error('An error was encountered while tailing the file')
+      error.code = 'ETAIL'
+      error.meta = {actual: err}
+      // Emitting on 'error' would bork the parent Readstream
+      // FIXME: error can be emitted here
+      this.emit('tail_error', error)
+    }
+    return
+  }
+
+  _read(size) {
```

Comment:
Is size unused on purpose?

Reply:
That's there because it's implementing node's `Readable` interface which has it as a param.  Their docs even say that it's really just info and doesn't need to be used, but I thought to include it for clarity.

https://nodejs.org/dist/latest-v12.x/docs/api/stream.html#stream_readable_read_size_1

> The size argument is advisory. For implementations where a "read" is a single operation that returns data can use the size argument to determine how much data to fetch. Other implementations may ignore this argument and simply provide data whenever it becomes available. There is no need to "wait" until size bytes are available before calling stream.push(chunk).

Reply:
That's...weird. But alrighty then.

Reply:
I'm not opposed to removing it.  Makes sense to I guess.

Reply:
Removed.